### PR TITLE
Added faster conversion method

### PR DESCRIPTION
**Pull request by snyball**
This method also doesn't hang Emacs, and can take advantage of multiple cores.

It does add a dependency on `fd` (I am conspiring to get it installed everywhere, but we can drop it for `find` if you don't want another dependency.)

This also makes the setup a bit easier, as the user doesn't have to bother with typing in their `$RUSTUP_HOME` and long default target name.

### CODE REVIEW

**Comment by samhedin** [comment]
Wow, that is *super* fast. Very impressive, and fixes the main issue I see with rustdoc at the moment.

>It does add a dependency on fd (I am conspiring to get it installed everywhere, but we can drop it for find if you don't want another dependency.)

I'm fine with it. I also prefer `fd` and I expect most people who want to use rustdoc already have it installed.

The default Pandoc org writer inserts a lot of empty newlines. After the old conversion `rustdoc-remove-whitespace` was run to clean this up. We should keep that functionality or add similar to `convert.sh`.

I also would like to retain `rustdoc-convert-current-package`. That way users can easily convert documentation specific to their current project.

I am not sure about the save location for the org files. There is a slight advantage to keeping it categorized under toolchain, but it also makes rg slower while searching, and it is not obvious where one would put the package-specific documentation as per the previous point. It should continue to respect `rustdoc-search-directory`, or some other user-specified variable.

Many of the files in std are redirects, and there is no need to convert them. They were filtered out previously with 
``` emacs-lisp
(< 10 (count-lines (point-min) ; redirect files are 10 lines long
                               (point-max)))
```
It's not a huge deal, but it would be nice to have similar functionality in the script.

In summary, the speed has increased massively but some functionality has been lost. Could we dump the script into `rustdoc.el`, and run it with `start-process`? This, along with some glue code, would restore most/all flexibility lost from switching from elisp to bash, make future packaging easier (End goal is that users should be able to install rustdoc through melpa), while still being insanely fast.

What do you think?

**Comment by snyball** [comment]
@samhedin 

Adapting those functions to call the `./convert.sh` script sounds like a very good idea, I'll do that.

Removing newlines should be easy to do with an `awk`/`sed` filter. With this excerpt from the converted `Vec` documentation, what newlines did the old solution remove?:

<details>

```
*** pub const fn new() -> Vec<T>

Constructs a new, empty =Vec<T>=.

The vector will not allocate until elements are pushed onto it.

#+BEGIN_SRC rust 
let mut vec: Vec<i32> = Vec::new();
#+END_SRC

*** pub fn with_capacity(capacity: usize) -> Vec<T>

Constructs a new, empty =Vec<T>= with the specified capacity.

The vector will be able to hold exactly =capacity= elements without
reallocating. If =capacity= is 0, the vector will not allocate.

It is important to note that although the returned vector has the
/capacity/ specified, the vector will have a zero /length/. For an
explanation of the difference between length and capacity, see /Capacity
and reallocation/.

#+BEGIN_SRC rust 
let mut vec = Vec::with_capacity(10);

// The vector contains no items, even though it has capacity for more
assert_eq!(vec.len(), 0);

// These are all done without reallocating...
for i in 0..10 {
    vec.push(i);
}

// ...but this may make the vector reallocate
vec.push(11);
#+END_SRC

*** pub fn into_raw_parts(self) -> (*mut T, usize, usize)

This is a nightly-only experimental API. (=vec_into_raw_parts= #65816)
new API

Decomposes a =Vec<T>= into its raw components.

Returns the raw pointer to the underlying data, the length of the vector
(in elements), and the allocated capacity of the data (in elements).
These are the same arguments in the same order as the arguments to
=from_raw_parts=.

After calling this function, the caller is responsible for the memory
previously managed by the =Vec=. The only way to do this is to convert
the raw pointer, length, and capacity back into a =Vec= with the
=from_raw_parts= function, allowing the destructor to perform the
cleanup.

#+BEGIN_SRC rust 
#![feature(vec_into_raw_parts)]
let v: Vec<i32> = vec![-1, 0, 1];

let (ptr, len, cap) = v.into_raw_parts();

let rebuilt = unsafe {
    // We can now make changes to the components, such as
    // transmuting the raw pointer to a compatible type.
    let ptr = ptr as *mut u32;

    Vec::from_raw_parts(ptr, len, cap)
};
assert_eq!(rebuilt, [4294967295, 0, 1]);
#+END_SRC
```

</details>

I think we might want to keep those redirects, but instead turn them into symlinks, because we might eventually want to turn the `<a href="...">` links into org links between the converted documents. That is, if those links are used internally, if they are solely there because of external references then they should be removed, but if that's the case then we can probably do the following:

1. Copy the doc-sets into something like `~/.local/share/emacs/rust-docsets/<crate>`
  - IMO It's probably better to not pollute the main docset directory anyway
2. Remove all the redirects using something like `rg -l "<p>Redirecting to <a href=\"[^\"]*\"" $path | xargs rm`
3. Run the `./convert.sh` script

Then we could probably bikeshed where to place the docsets, if we both agree they should be moved from `$RUSTUP_HOME`.

**Comment by snyball** [comment]
@samhedin This script runs in 6.1 seconds on my computer, so filtering the old way is definitely doable:

```bash
#!/usr/bin/env bash

DOC_PATH="$(dirname "$(rustup doc --std --path)")"

for file in $(fd . -ehtml "$DOC_PATH"); do
    # Use head first to avoid reading the whole file
    if [ $(head -n 11 "$file" | wc -l) != 10 ]; then
        echo "$file"
    fi
done
```

**Comment by snyball** [comment]
@samhedin `rg -l "<p>Redirecting to <a href=\"[^\"]*\""` really seems to be the way to go, it only takes 108ms and should be way more reliable (but obviously not perfect.)

**Comment by samhedin** [comment]
>Adapting those functions to call the ./convert.sh script sounds like a very good idea, I'll do that.

Cool! This will really up the quality of rustdoc.

> Removing newlines should be easy to do with an `awk`/`sed` filter. With this excerpt from the converted `Vec` documentation, what newlines did the old solution remove?:

The old filter removed all newlines except those followed by a header (defined as a line starting with *).
> ```
> *** pub const fn new() -> Vec<T>
> Constructs a new, empty =Vec<T>=.
> The vector will not allocate until elements are pushed onto it.
> #+BEGIN_SRC rust 
> let mut vec: Vec<i32> = Vec::new();
> #+END_SRC
> 
> *** pub fn with_capacity(capacity: usize) -> Vec<T>
> Constructs a new, empty =Vec<T>= with the specified capacity.
> The vector will be able to hold exactly =capacity= elements without...
and so on. The end result is not set in stone. It was a quick cleanup because I found all the whitespace distracting and this was an easy way to make it more compact.

>Copy the doc-sets into something like ~/.local/share/emacs/rust-docsets/<crate>

Do you mean the .HTML docs? No need to copy them or mix them with .org files.
The user should be able to specify a directory and then have `output.sh` place the converted org files in `rustdoc-search-directory`. The subdirectory structure can be kept by taking the relative file name.
I might be misunderstanding something here, so please clarify if I just said something strange.

>rg -vl [...] really seems to be the way to go, it only takes 108ms and should be way more reliable (but obviously not perfect.)

Great! As far as changing the links we want to keep, it might be easier to do the conversion inside the pandoc filter, though I am not sure. Either way, that is probably a later issue.

**Comment by snyball** [comment]
@samhedin `convert.sh` is now a bit faster, as it will skip the redirects, thankfully I found the `--ignore-file` option of `fd` which made this much easier (and without adding anything to the run-time.)

It also means that it doesn't have to do a silly "copy, then remove" process.

**Comment by snyball** [comment]
@samhedin As for the newline issue, is it possible to handle this in the Lua filter?

**Comment by snyball** [comment]
@samhedin I think it would have to be handled by the Lua script, if at all, as the simple fix would crunch these longer paragraphs in such a way that they just become harder to read (this is subjective of course):

### Original

<details>

```
Creates a =Vec<T>= directly from the raw components of another vector.

This is highly unsafe, due to the number of invariants that aren't
checked:

- =ptr= needs to have been previously allocated via =String=/=Vec<T>=
  (at least, it's highly likely to be incorrect if it wasn't).
- =T= needs to have the same size and alignment as what =ptr= was
  allocated with. (=T= having a less strict alignment is not sufficient,
  the alignment really needs to be equal to satsify the =dealloc=
  requirement that memory must be allocated and deallocated with the
  same layout.)
- =length= needs to be less than or equal to =capacity=.
- =capacity= needs to be the capacity that the pointer was allocated
  with.

Violating these may cause problems like corrupting the allocator's
internal data structures. For example it is *not* safe to build a
=Vec<u8>= from a pointer to a C =char= array with length =size_t=. It's
also not safe to build one from a =Vec<u16>= and its length, because the
allocator cares about the alignment, and these two types have different
alignments. The buffer was allocated with alignment 2 (for =u16=), but
after turning it into a =Vec<u8>= it'll be deallocated with alignment 1.

The ownership of =ptr= is effectively transferred to the =Vec<T>= which
may then deallocate, reallocate or change the contents of memory pointed
to by the pointer at will. Ensure that nothing else uses the pointer
after calling this function.
```

</details>

### Crunched

<details>

```
Creates a =Vec<T>= directly from the raw components of another vector.
This is highly unsafe, due to the number of invariants that aren't
checked:
- =ptr= needs to have been previously allocated via =String=/=Vec<T>=
  (at least, it's highly likely to be incorrect if it wasn't).
- =T= needs to have the same size and alignment as what =ptr= was
  allocated with. (=T= having a less strict alignment is not sufficient,
  the alignment really needs to be equal to satsify the =dealloc=
  requirement that memory must be allocated and deallocated with the
  same layout.)
- =length= needs to be less than or equal to =capacity=.
- =capacity= needs to be the capacity that the pointer was allocated
  with.
Violating these may cause problems like corrupting the allocator's
internal data structures. For example it is *not* safe to build a
=Vec<u8>= from a pointer to a C =char= array with length =size_t=. It's
also not safe to build one from a =Vec<u16>= and its length, because the
allocator cares about the alignment, and these two types have different
alignments. The buffer was allocated with alignment 2 (for =u16=), but
after turning it into a =Vec<u8>= it'll be deallocated with alignment 1.
The ownership of =ptr= is effectively transferred to the =Vec<T>= which
may then deallocate, reallocate or change the contents of memory pointed
to by the pointer at will. Ensure that nothing else uses the pointer
after calling this function.
```

</details>

**Comment by snyball** [comment]
@samhedin Should `rustdoc-to-org` be based on manually searching, or should it also support getting information with `lsp` and then searching for that?

I'm thinking that most people writing mid-large size Rust projects in Emacs are probably using LSP, I'm thinking of hacking an org-mode display into `lsp-describe-thing-at-point`, but I'll attempt that in a different pull request. This work could then be merged into `rustic-mode`.

**Comment by samhedin** [comment]
> convert.sh is now a bit faster, as it will skip the redirects, thankfully I found the --ignore-file option of fd which made this much easier (and without adding anything to the run-time.)

Sick!

> @samhedin As for the newline issue, is it possible to handle this in the Lua filter?

The filter function does not include the newlines, the issue arises when Pandoc takes the filtered data and converts it to org. Luckily, it's about to get much better: https://github.com/jgm/pandoc/issues/6554
Initially, I thought that this change in Pandoc would be nice but not sufficient, but I am starting to think that it might be just right. 
We can leave this for now - going by the release frequency of Pandoc this change should be released soon enough.

> With std being default for library, the default for target is whatever rustup show says it is.

This solution is convenient for converting std, but doesn't this make it impossible to convert package-specific documentation?
``` bash
DOC_PATH="$RUSTUP_HOME/toolchains/$TARGET/share/doc/rust/html/$LIBRARY"
```
If it's just changed to take a full directory path (we can query the user for it in emacs to make it easy to enter), it can be used for anything.
``` bash
DOC_PATH="$LIBRARY" # Something like this
```

An alternative solution is fine too, but it should be straightforward to generate any folder with documentation. When that's done, I think this is pretty much ready to be merged. There is some elisp to write to call and interact the script, but it should not be too bad. If you want to you are very welcome to do it, otherwise, I can take care of it.

>Should rustdoc-to-org be based on manually searching, or should it also support getting information with lsp and then searching for that?
I started this project because I was using LSP, expected there to be thorough documentation when using `lsp-describe-thing-at-point`, and finding out that there wasn't (and this seems quite far down/nonexistent on the priority list of rust analyzer/rustdoc). So I'm all for integrating with LSP. I even looked at it a bit previously but decided to leave it for later. 

>EDIT: Or maybe merged into lsp-mode if made sufficiently generic, every language mode could be supplied with a custom pandoc filter to generate .org files.

Yeah, this would be super cool. This project does a similar thing (and was an inspiration for rustdoc), but is further along in development: https://github.com/danielmartin/swift-helpful

**Comment by samhedin** [comment]
I believe it is possible to make a pure elisp solution almost as fast as the current one. I worked on an async one previously but abandoned it because it kept throwing a "too many filedescriptors open" error at me. Maybe this can be solved, in which case such a solution would be preferable.

**Comment by snyball** [comment]
@samhedin Yeah, the script is currently only for converting documentation in the system directories, as a test.

I'm going to write the elisp-code tomorrow, I'm thinking it should enumerate the libraries in the current cargo project (rustic-mode mostly has that figured out I think) and just start converting all of it. Now that the process doesn't take too long I don't know if people need fine-grained control over which specific libraries they want to convert. I'll cache everything in `$HOME/.local/share/...` so that users won't have to redo the conversion if they use the same library in many crates.

Later on there should probably also be an `mtime` check on the documentation and the `org` output to see if they need to be re-generated.

**Comment by snyball** [comment]
@samhedin I think it's pretty much done now.

The last thing to be done is to just change `rustdoc-source-user` to `samhedin`.

File: rustdoc.el [hunk 1]
```diff
@@ -76,103 +96,68 @@ This is useful if you want to search for the name of a struct, enum or trait."
...
-      (if (with-temp-buffer
-            (insert-file-contents file)
-            (< 10 (count-lines (point-min)
-                               (point-max))))
-          (progn
-            (rustdoc-convert-file dir file))
-        (message "Ignoring conversion of %s as it is probably a redirect"
-                 file)))
-    (message "Batch conversion done!")))
+    (helm-ag search-directory (concat regex search-term))))
```
Comment (samhedin) [inline rustdoc.el:107]: Will this search both std and package-specific documentation?

File: rustdoc.el [hunk 2]
```diff
@@ -76,103 +96,68 @@ This is useful if you want to search for the name of a struct, enum or trait."
...
+     rustdoc-convert-prog
+     finish-func
+     docs-src
+     docs-dst)))
 
 
 ;;;###autoload
-(defun rustdoc-convert-current-package ()
-  "Generate and convert the documentation for the current rust package."
+(defun rustdoc-setup ()
```
Comment (samhedin) [inline rustdoc.el:151]: Running this returns "emacs: /home/sam/.local/bin/rustdoc-to-org-convert.sh: Exec format error"

File: rustdoc.el (hunk 1 above)
Comment (samhedin) [reply rustdoc.el:107]: Oh I see now. It's creating a symlink to the std directory if you have package specific documentation?
Smart. What happens if you do not have any package specific documentation?
(Asking since I can't run it at the moment, see below)

File: rustdoc.el (hunk 2 above)
Comment (snyball) [reply rustdoc.el:151]: Oh, did you try temporarily changing `rustdoc-source-user` to `snyball` then running the `setup` function to download the resources?

File: rustdoc.el (hunk 1 above)
Comment (snyball) [reply rustdoc.el:107]: Fixed a bug with the symlink as of 8197f43, if there is no package specific documentation, it will just search `std` via that symlink.

File: rustdoc.el (hunk 2 above)
Comment (snyball) [reply rustdoc.el:153]: What does `/home/sam/.local/bin/rustdoc-to-org-convert.sh` look like?

Comment (samhedin) [reply rustdoc.el:153]: That did it, thank you!

File: rustdoc.el [hunk 3]
```diff
@@ -76,103 +96,70 @@ This is useful if you want to search for the name of a struct, enum or trait."
...
-                  (file-truename rustdoc-lua-filter)
-                  "-o"
-                  (file-truename outputfile))
-    (rustdoc-remove-whitespace outputfile)))
+(if (< emacs-major-version 27)
+    (defun rustdoc--xdg-data-home ()
+      (or (getenv "XDG_DATA_HOME")
+          (concat (file-name-as-directory (getenv "HOME"))
+                  ".local/share")))
+  (fset 'rustdoc--xdg-data-home 'xdg-data-home))
```
[the remark this answers is not in the document]
Comment (snyball) [reply rustdoc.el]: Done, but I'm only doing `(require 'xdg)` if `emacs-major-version` is greater than 26.

File: rustdoc.el [hunk 4]
```diff
@@ -76,103 +96,70 @@ This is useful if you want to search for the name of a struct, enum or trait."
...
+         ;;        strings out of cargo, or just parse the Cargo.toml file, but
+         ;;        then we'd have to review different parsing solutions.
+         ;;
+         ;;        For now, I think this is an ok solution.
+         (docs-dst (concat (file-name-as-directory proj)
+                           rustdoc-local-directory))
+         (finish-func (lambda (p)
+                        (message (format "Finished converting docs for: %s" proj)))))
+    (make-directory (concat (file-name-as-directory proj)
+                            rustdoc-local-directory))
```
[the remark this answers is not in the document]
Comment (snyball) [reply rustdoc.el]: Done in 4e4b84f

**Comment by samhedin** [comment]
Super great job! Rustdoc is much better now.
